File: lib/mix/tasks/tunneld.agent.ex
defmodule Mix.Tasks.Tunneld.Agent do
  @shortdoc "tunneld CLI - thin HTTP wrapper over the agent API"
  @moduledoc """
  Thin CLI over the tunneld agent HTTP API (no policy; the API is the product).

      mix tunneld.agent <command> [args...]

  Commands:
    machines                          GET  /agent/machines
    machine <id>                      GET  /agent/machines/:id
    listeners <id>                    GET  /agent/machines/:id/listeners
    probe <id>                        POST /agent/machines/:id/probe  (202 job)
    exec <id> <cmd>                   POST /agent/machines/:id/exec   (202 job)
    job <id>                          GET  /agent/jobs/:id
    resources                         GET  /agent/resources
    add-resource <name> <pool...>     POST /agent/resources
    rm-resource <id>                  DELETE /agent/resources/:id

  Env:
    TUNNELD_URL  (default http://10.0.0.1)
    TUNNELD_TOKEN (required; scoped tnld_ bearer token)
  """
  use Mix.Task

  @impl true
  def run(args) do
    base = System.get_env("TUNNELD_URL") || "http://10.0.0.1"
    token = System.get_env("TUNNELD_TOKEN")

    if is_nil(token) do
      IO.puts("TUNNELD_TOKEN required (issue via: mix tunneld.issue_token machines:read,resources:write)")
      exit({:shutdown, 1})
    end

    {result, _code} = dispatch(base, token, args)

    IO.puts(Jason.encode!(result, pretty: true))

    if is_map(result) and Map.get(result, "error") do
      exit({:shutdown, 1})
    end
  end

  defp dispatch(base, token, [cmd | rest]) do
    case cmd do
      "machines" -> api(base, token, :get, "/api/v1/agent/machines")
      "machine" -> api(base, token, :get, "/api/v1/agent/machines/#{hd(rest)}")
      "listeners" -> api(base, token, :get, "/api/v1/agent/machines/#{hd(rest)}/listeners")
      "probe" -> api(base, token, :post, "/api/v1/agent/machines/#{hd(rest)}/probe")
      "exec" -> api(base, token, :post, "/api/v1/agent/machines/#{hd(rest)}/exec", %{"cmd" => Enum.join(tl(rest), " ")})
      "job" -> api(base, token, :get, "/api/v1/agent/jobs/#{hd(rest)}")
      "resources" -> api(base, token, :get, "/api/v1/agent/resources")
      "add-resource" ->
        [name | pool] = rest
        api(base, token, :post, "/api/v1/agent/resources", %{"name" => name, "pool" => pool})
      "rm-resource" -> api(base, token, :delete, "/api/v1/agent/resources/#{hd(rest)}")
      _ -> {{%{"error" => "unknown command: #{cmd}"}, :json}}
    end
  end

  defp dispatch(_base, _token, []), do: {{%{"error" => "usage: tunneld.agent <command> [args]"}, :json}}

  defp api(base, token, method, path, body \\ nil) do
    url = base <> path
    headers = [{"authorization", "Bearer #{token}"}, {"content-type", "application/json"}]

    result =
      case method do
        :get -> HTTPoison.get(url, headers)
        :post -> HTTPoison.post(url, Jason.encode!(body || %{}), headers)
        :delete -> HTTPoison.delete(url, headers)
      end

    case result do
      {:ok, %HTTPoison.Response{status_code: s, body: b}} ->
        case Jason.decode(b) do
          {:ok, decoded} -> {decoded, :json}
          _ -> {{%{"status" => s, "body" => b}}, :json}
        end

      {:error, reason} ->
        {{%{"error" => inspect(reason)}, :json}}
    end
  end
end

File: skill/tunneld_agent.py
"""Thin tunneld agent API client (no policy). The HTTP API is the product."""
import os
import urllib.request
import urllib.error
import json


class TunneldError(RuntimeError):
    pass


class TunneldClient:
    def __init__(self, url=None, token=None, timeout=30):
        self.url = (url or os.environ.get("TUNNELD_URL") or "http://10.0.0.1").rstrip("/")
        self.token = token or os.environ.get("TUNNELD_TOKEN")
        if not self.token:
            raise TunneldError("TUNNELD_TOKEN required (issue via: mix tunneld.issue_token ...)")
        self.timeout = timeout

    def _request(self, method, path, body=None):
        req = urllib.request.Request(self.url + path, method=method)
        req.add_header("Authorization", "Bearer " + self.token)
        req.add_header("Content-Type", "application/json")
        data = json.dumps(body).encode() if body is not None else None
        try:
            with urllib.request.urlopen(req, data=data, timeout=self.timeout) as resp:
                raw = resp.read()
                return json.loads(raw) if raw else {}
        except urllib.error.HTTPError as e:
            msg = e.read().decode(errors="replace")
            raise TunneldError(f"HTTP {e.code}: {msg}") from e

    def machines(self):
        return self._request("GET", "/api/v1/agent/machines")

    def machine(self, id):
        return self._request("GET", f"/api/v1/agent/machines/{id}")

    def listeners(self, id):
        return self._request("GET", f"/api/v1/agent/machines/{id}/listeners")

    def probe(self, id):
        return self._request("POST", f"/api/v1/agent/machines/{id}/probe")

    def exec(self, id, cmd):
        return self._request("POST", f"/api/v1/agent/machines/{id}/exec", {"cmd": cmd})

    def jobs(self, id):
        return self._request("GET", f"/api/v1/agent/jobs/{id}")

    def resources(self):
        return self._request("GET", "/api/v1/agent/resources")

    def add_resource(self, name, pool):
        return self._request("POST", "/api/v1/agent/resources", {"name": name, "pool": pool})

    def rm_resource(self, id):
        return self._request("DELETE", f"/api/v1/agent/resources/{id}")
